### ISSUE TITLE
Linker warning on OS X

### ISSUE THREAD

**Issue by beemaster** [c42f/tinyformat#25]
I have a linker warning on OS X:

ld: warning: direct access in tinyformat::detail::FormatArg::FormatArg<unsigned long>(unsigned long const&) to global weak symbol void tinyformat::detail::FormatArg::formatImpl<unsigned long>(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, char const*, int, void const*) means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.

Do you have an idea how to fix it?

**Comment by c42f** [comment]
The following two stack overflow questions seem to be relevant:

http://stackoverflow.com/questions/8685045/xcode-with-boost-linkerid-warning-about-visibility-settings
http://stackoverflow.com/questions/9894961/strange-warnings-from-the-linker-ld

From those posts I guess that this problem might be due to using tinyformat in two different projects with different default symbol visibilities, and then linking them together?  In that case it's not a tinyformat problem per se - you should check your settings for "Symbols Hidden by Default" in xcode for both projects.

Can you tell me the compiler and version?  I'm assuming it's gcc based on the error message.

**Comment by beemaster** [comment]
Compiler is Apple LLVM version 6.0 (clang-600.0.57) (based on LLVM 3.5svn)
Most likely you are right, the problem is in project visibility settings.

I am using following setup:
1) Static library that uses tinyformat. It has Symbols Hidden by Default set to NO. This is default XCode setting for static libraries.
2) Application that uses both static library and tinyformat. It has Symbols Hidden by Default set to YES. This is default XCode setting for applications.

Since this is standard setup, I am not sure I want to change visibility settings. Do you think there is a possibility to change it in tinyformat?

**Comment by c42f** [comment]
There might be a workaround we could do in the source code.  Perhaps it's as simple as adding `__attribute__((visibility("hidden")))` attributes in a couple of places in the private `FormatArg` thunks?  This would need to be behind a macro (`TINYFORMAT_HIDDEN`?) so that it doesn't mess up MSVC compatibility.

I can't reproduce this problem on linux with g++-4.8.2 or clang++-3.5, and I don't have ready access to OSX, but maybe you can get a cut down example going.  Here's what I did:

Makefile:
```make
CXX=clang++-3.5

INCLUDE=-I$(HOME)/dev/tinyformat

all: main

foo.o: Makefile foo.cpp foo.h
	$(CXX) -Wall $(INCLUDE) -c -fvisibility=default foo.cpp

test.o: Makefile test.cpp foo.h
	$(CXX) -Wall $(INCLUDE) -c -fvisibility=hidden test.cpp

main: Makefile test.o foo.o
	$(CXX) -Wall test.o foo.o -o main
```

foo.h:
```C++
#pragma once

void foo();
```

foo.cpp:
```C++
#include <tinyformat.h>
#include "foo.h"

void foo()
{
    tfm::printf("%d\n", 2);
}
```

test.cpp:
```C++
#include <tinyformat.h>

#include "foo.h"

int main()
{
    foo();
    tfm::printf("%d\n", 1);
    return 0;
}
```

**Comment by beemaster** [comment]
I compiled your example and was able to reproduce the warning. The fix that you suggested worked well, thanks! I created a pull request https://github.com/c42f/tinyformat/pull/26

**Comment by c42f** [comment]
Fixed by #26
